### ISSUE TITLE
Sometimes sops store broken data key (invalid key size).

### ISSUE THREAD

**Issue by azhi** [mozilla/sops#278]
After a few key rotations sops can leave encrypted file in a broken state.

Steps to reproduce:
1) Generate new encrypted file with single PGP key:
`sops -pgp <fingerprint> tmp.yml`
2) Launch a loop that constantly rotates data key: for bash
`while [[ $? == 0 ]]; do; sops -r -i tmp.yml; done`
3) Wait until the loop fails with error:

```
[PGP]	 INFO[0000] Decryption succeeded                          fingerprint=<fingerprint>
[SOPS]	 INFO[0000] Data key recovered successfully              
Error decrypting tree: Error walking tree: Could not decrypt value: crypto/aes: invalid key size 31
```

It usually takes 5-10 cycles to break the file.

This behaviour is observed in version 3.0.0 (in linux version downloaded from [releases](https://github.com/mozilla/sops/releases/download/3.0.0/sops-3.0.0.linux), as well as in darwin version installed by `brew install sops`).

PGP agent:
```
$ gpg --version
gpg (GnuPG) 2.1.21
libgcrypt 1.7.7
```

Example of decrypted data key from broken file:

```
$ cat tmp.yml | ruby -ryaml -e 'puts YAML.load(STDIN.read)["sops"]["pgp"][0]["enc"]' | gpg -d | hexdump -C
00000000  9c 74 86 a4 54 81 72 e1  92 9f f5 80 1d d0 c4 92  |.t..T.r.........|
00000010  9e 39 b7 dc 00 93 5f d7  3e 15 3f 78 13 69 3d     |.9...._.>.?x.i=|
0000001f
```

**Comment by autrilla** [comment]
I managed to reproduce this. As far as I can tell, this is a problem that only affects PGP. The cause appears to be PGP encryption not working in some cases. I've added some debug logging, and added a check that whenever something is encrypted using PGP, it decrypts the result and checks that it is indeed the same as the input.

Here's a log of it failing:

```
[PGP]	 time="2017-11-27T19:19:48+01:00" level=info msg="Decryption succeeded" fingerprint=1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
[SOPS]	 time="2017-11-27T19:19:48+01:00" level=info msg="Data key recovered successfully"
[SOPS]	 time="2017-11-27T19:19:48+01:00" level=info msg="Data key len: 32"
[CMD]	 time="2017-11-27T19:19:48+01:00" level=info msg="Old data key: 39CED5667311A2DEF7E6BA907CB027BCBF2E083294B18569186408DB4BCCF34E"
[PGP]	 time="2017-11-27T19:19:49+01:00" level=info msg="Encryption succeeded" fingerprint=1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
[PGP]	 time="2017-11-27T19:19:49+01:00" level=info msg="Asked to encrypt BDDB32F4DF0C4CA952CE9FB3EB2290E7983EDDD7FC85AC0D646F90F3D11BDB2E"
[PGP]	 time="2017-11-27T19:19:49+01:00" level=info msg="Encrypted to -----BEGIN PGP MESSAGE-----\n\nwYwDEEVDpnzXnMABBACNQ/aMNN3wZ6RQu5wT+EXqnDdBRDx7DCUrPj5T/LUnh7bN\n+vivQFWU+40HcDu69R5Z23Ch1mvfSJtxAGibEqOnPFmWOztTndvZGcjDWXe7Uu8K\nvq3DslYgAmiDhsbwcfDYtdqwX26XpMj+qiY3FHGelWMWUNjo7aUtriW6RrNrw9Lg\nAeS2w034qHIu3/kvkmGbULzh4S3X4E7gUuH2heC04hXdSeXgR+USmh86MZynpF0/\nG88oJnNypFe+AkCOfdGqCCLr1LvFhOAK5Cr+10cPx+tcCo7XEJ2GhCviZKTAdOGo\nuwA=\n=TwTb\n-----END PGP MESSAGE-----"
[PGP]	 time="2017-11-27T19:19:49+01:00" level=info msg="Decryption succeeded" fingerprint=1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
[PGP]	 time="2017-11-27T19:19:49+01:00" level=info msg="This decrypts to BDDB32F4DF0C4CA952CE9FB3EB2290E7983EDDD7FC85AC646F90F3D11BDB2E"
```

So we're attempting to reencrypt the file with a new key (hex encoded), `BDDB32F4DF0C4CA952CE9FB3EB2290E7983EDDD7FC85AC0D646F90F3D11BDB2E`. We then encrypt this with PGP, obtaining a ciphertext. If we decrypt this ciphertext with PGP, we get `BDDB32F4DF0C4CA952CE9FB3EB2290E7983EDDD7FC85AC646F90F3D11BDB2E`. Here they are one above the other so it's easy to tell they're different:

```
New generated data key:       BDDB32F4DF0C4CA952CE9FB3EB2290E7983EDDD7FC85AC0D646F90F3D11BDB2E
Returned from PGP round-trip: BDDB32F4DF0C4CA952CE9FB3EB2290E7983EDDD7FC85AC646F90F3D11BDB2E
```

**Comment by autrilla** [comment]
What's interesting is that the automated quickcheck-like test we have for this does fail for me consistently:
```
❯ go test go.mozilla.org/sops/pgp
[PGP]	 time="2017-11-27T19:33:05+01:00" level=info msg="Encryption succeeded" fingerprint=1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
[PGP]	 time="2017-11-27T19:33:05+01:00" level=info msg="Decryption succeeded" fingerprint=1022470DE3F0BC54BC6AB62DE05550BC07FB1A0A
--- FAIL: TestPGP (1.79s)
	keysource_test.go:29: #1: failed on input []byte{0x3b, 0xc6, 0xd4, 0x87, 0x63, 0xfa, 0x2, 0x43, 0x85, 0xc4, 0x45, 0x92, 0x42, 0x93, 0x7, 0x5c, 0x9c, 0x9c, 0xd, 0xba, 0x1b, 0xe8, 0xc1, 0x3c, 0xb2, 0xd7, 0x12, 0xf7, 0x16, 0xee, 0x86, 0x7f, 0x57, 0xb7, 0xed, 0x5, 0xa5, 0x2e, 0x25}
FAIL
FAIL	go.mozilla.org/sops/pgp	1.963s
```

, but it seems to [work fine on CI](https://travis-ci.org/mozilla/sops/builds/304656807#L664).

**Comment by azhi** [comment]
@autrilla yes, it does.

I've digged into this a little, and it seems the issue is with golang's `crypto/openpgp` encrypting.

In all tests sops encodes through `crypto/openpgp` succesfully, then fails to decode through `crypto/openpgp` with `Reading PGP message failed: openpgp: incorrect key` and fallbacks to binary gpg.

I've found an issue https://github.com/golang/go/issues/13409, which says that GPG treats input data as binary data, and golang's `crypto/openpgp` treats input data as text data.

Fix is fairly simple, I'll make a PR in a few minutes.

**Comment by azhi** [comment]
> In all tests sops encodes through `crypto/openpgp` succesfully, then **fails to decode through `crypto/openpgp` with `Reading PGP message failed: openpgp: incorrect key`**

@autrilla Oh, and btw, this is probably the reason it works in travis and for other people (they can decode through `crypto/openpgp`), but not for you and me. I didn't look into this issue at all. After my fix in the PR it should not matter, but it might be worth looking into anyway.

**Comment by autrilla** [comment]
Actually, I can't reproduce this anymore. While debugging it I tried switching GPG and Go versions, and at some point I apparently fixed this issue for me, but I don't know exactly how.

Yeah, in my case because I have a Yubikey key on my keyring (and also use GPG2), I'm forced to use the `gpg` binary. But yeah, this apparently fixed itself for me, somehow.

**Comment by autrilla** [comment]
It's worth noting that the aforementioned tests did pass for me a few months ago, when I was working on SOPS more regularly -- so this might have been caused by a change to GPG or the `crypto/openpgp` package.

**Comment by azhi** [comment]
It is still failing for me.

Based on linked golang issue, it fails any time we encrypt with openpgp and then decrypt with gpg. If we encrypt and decrypt with openpgp - everything is fine. If we encrypt and decrypt with gpg - everything is fine. Only when we feed output of openpgp into gpg we get the issue.

Also you might just be lucky not to have 0x0D in your test data :).

**Comment by autrilla** [comment]
Yeah, it's clear to me now what the culprit was. Thanks a lot of finding this!

No, the test that was failing is quicktest-like, so it randomly tries a bunch of inputs, and it failed every single time I ran it while I was able to reproduce the issue, as it tries a very large number of inputs, and it's extremely unlikely they wouldn't contain 0x0d. I definitely did something that fixed it for me, but I still have no idea what.
